cli/lsp/sources.rs

```diff
@@ -0,0 +1,382 @@
+
+use super::analysis;
+use super::text;
+
+use crate::file_fetcher::get_source_from_bytes;
+use crate::file_fetcher::map_content_type;
+use crate::http_cache;
+use crate::http_cache::HttpCache;
+use crate::media_type::MediaType;
+use crate::text_encoding;
+
+use deno_core::serde_json;
+use deno_core::ModuleSpecifier;
+use std::collections::HashMap;
+use std::fs;
+use std::path::Path;
+use std::path::PathBuf;
+use std::time::SystemTime;
+
+#[derive(Debug, Clone, Default)]
+struct Metadata {
+  dependencies: Option<HashMap<String, analysis::Dependency>>,
+  maybe_types: Option<analysis::ResolvedImport>,
+  media_type: MediaType,
+  source: String,
+  version: String,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct Sources {
+  http_cache: HttpCache,
+  metadata: HashMap<ModuleSpecifier, Metadata>,
+  redirects: HashMap<ModuleSpecifier, ModuleSpecifier>,
+  remotes: HashMap<ModuleSpecifier, PathBuf>,
+}
+
+impl Sources {
+  pub fn new(location: &Path) -> Self {
+    Self {
+      http_cache: HttpCache::new(location),
+      ..Default::default()
+    }
+  }
+
+  pub fn contains(&mut self, specifier: &ModuleSpecifier) -> bool {
+    if let Some(specifier) = self.resolve_specifier(specifier) {
+      if self.get_metadata(&specifier).is_some() {
+        return true;
+      }
+    }
+    false
+  }
+
+  pub fn get_length(&mut self, specifier: &ModuleSpecifier) -> Option<usize> {
+    if let Some(specifier) = self.resolve_specifier(specifier) {
+      if let Some(metadata) = self.get_metadata(&specifier) {
+        return Some(metadata.source.chars().count());
+      }
+    }
+    None
+  }
+
+  pub fn get_line_index(
+    &mut self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<Vec<u32>> {
+    if let Some(specifier) = self.resolve_specifier(specifier) {
+      if let Some(metadata) = self.get_metadata(&specifier) {
+        return Some(text::index_lines(&metadata.source));
+      }
+    }
+    None
+  }
+
+  pub fn get_media_type(
+    &mut self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<MediaType> {
+    if let Some(specifier) = self.resolve_specifier(specifier) {
+      if let Some(metadata) = self.get_metadata(&specifier) {
+        return Some(metadata.media_type);
+      }
+    }
+    None
+  }
+
+  fn get_metadata(&mut self, specifier: &ModuleSpecifier) -> Option<Metadata> {
+    if let Some(metadata) = self.metadata.get(specifier).cloned() {
+      if let Some(current_version) = self.get_script_version(specifier) {
+        if metadata.version == current_version {
+          return Some(metadata);
+        }
+      }
+    }
+    if let Some(version) = self.get_script_version(specifier) {
+      if let Some(path) = self.get_path(specifier) {
+        if let Ok(bytes) = fs::read(path) {
+          if specifier.as_url().scheme() == "file" {
+            let charset = text_encoding::detect_charset(&bytes).to_string();
+            if let Ok(source) = get_source_from_bytes(bytes, Some(charset)) {
```

Comment:
Using `?` instead of `if let Some()` should make this function a lot less indented.

Reply:
No, because I can't/don't want to return a result.  These methods can't error.

Reply:
Okay, but you can use `?` with options as well - encountering `None` will short-circuit just like in case of `Err`

Reply:
🤯  I was using the `?` operator without understanding what it actually did... 😭   I cleaned up where it was obvious.